cmd/tealdbg/debugger_test.go

```diff
@@ -185,3 +191,91 @@ func TestSession(t *testing.T) {
 	require.NotEmpty(t, name)
 	require.Greater(t, len(data), 0)
 }
+
+// Tests control functions for stepping over subroutines and checks
+// that call stack is inspected correctly.
+func TestCallStackControl(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	s := createSessionFromSource(t, "#pragma version %d\nlab1:\nint 1\ncallsub lab1\ndup\n+\n")
+	s.callStack = []logic.CallFrame{
+		{FrameLine: 2, LabelName: "lab1"},
+	}
+	// Check for step over on a callsub line
+	s.line.Store(3)
```

Comment:
@algochoi For my understanding:  Had you considered writing the test _with_ a `Debugger`?

* I'm in the middle of understanding the added tests.  I'm gauging if the question's been considered because I _suspect_ the choice impacts a number of other decisions. 
* From what I've understood so far, my primary concern is:  The test modifies `Session` state without using publicly available facilities.  I'm concerned a test is liable to sequence steps in a way that's incompatible with public APIs.

Reply:
I guess the main reason I wrote these tests was because it was compatible to the existing unit tests.

Most of the debugger control functions are issued by `cdtSession` through the `Control` interface:

https://github.com/algorand/go-algorand/blob/1d8bd1bd79178bc1c76cd1afb41da9d2e639a806/cmd/tealdbg/cdtSession.go#L39 

I can see your concern, and I think these tests mostly test public APIs (except for setting `s.line` and `s.callstack`, which we have to manually set for these tests due to no access to the evaluator), but if you see something specific wrong, I can try to address it.

Another concern you might be having is that we're not testing end-to-end behavior (we don't actually get real-time state from the evaluator). That's a fair point, and we could write a more end-to-end testing framework for tealdbg in the future. I fear that the diff for that might make this PR much bigger than we want.

Reply:
@algochoi Thanks for the response - it addresses some of my concerns.  I tried to outline what I'm thinking below.  

I think there's a lot happening here.  I leave the note to illustrate my thinking.  I'll schedule some time to discuss because I think a verbal will move faster.  We can summarize outcomes afterwards.

* I think these tests ought to step through a problem via `Control` using `Debugger`.  `Debugger` contains the update logic (`Update`) and a reference to `session` (implements `Control`). 
* From what I can tell, the proposed tests do _not_ have a CDT dependency.  I don't think I'm proposing an _end-to-end_ test.  
* Regarding correctness, I think manually changing state (e.g. line numbers) is too likely to produce unrepresentable states.  
  * For example, I think https://github.com/algorand/go-algorand/pull/3653#discussion_r834600725 illustrates a broken state.  Getting to line 4 with a non-empty call stack _should not_ be possible.
  * If my understanding is accurate, then I'm left feeling like I can't rely on these tests.  
* I understand the request extends the PR. 
  * I am open to finding ways to merge what's open and use successive PRs to keep scope tighter.
  * My principal concern is agreeing on the approach for correctness.

Reply:
RE:  off line discussion

I think I am beginning to understand your concerns and I'm working on a stronger test for recursion - we can loop back after I push some changes I am working on.

Reply:
I pushed some changes - they are actually mostly minor and might not address your initial problem: I'll explain my reasoning below. 

I'm getting more convinced that it will turn into an end-to-end test if we really want to test recursive calls or any other runtime conditions without calling the actual evaluator. My initial thoughts on this unit test is that it should not depend on a particular implementation on the node's evaluator. Since the evaluator is responsible for updating the callstack and the line numbers, then I think it's okay to set these ourselves in the unit tests. I think the current debugger is not supposed to be exposed to how the TEAL program is supposed to run anyway. 

I guess you could propose to write our own evaluator (probably a subset of it; writing the whole thing would take a while 😅 ) in the debugger to simulate recursive calls and alike, but I still think it couples our debugger with our current implementation of the evaluator. 

Let me know if this makes sense and if you have any other concerns.

Reply:
@algochoi Gotcha - A few points:
* I'll need more time to consider these thoughts.  I'll do some homework and circle back.
* I opened https://github.com/algochoi/go-algorand/pull/2 to make the test cases independent.  Can you take a look?  
  * Since the PR is opened against your fork, I think you'll need to approve the workflow run:  https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks.   
  * When I began reviewing again, I struggled to understand the initial state for each test case.  Splitting the cases ensures each scenario acts in isolation.  
  * If you think it's an improvement, I'd like to merge as soon as is reasonably possible.  The PR's changes are liable to result in conflicts.

Reply:
@algochoi I apologize for the long response time.  I'm good to resolve the thread.  Leaving details below if interested.

* Thanks for merging in https://github.com/algochoi/go-algorand/pull/2.
* I agree attempting to use the _real_ evaluator moves from unit to integration test territory. 
  * I attempted to write such a test locally and see there's a number of moving parts.  Given the desire to get something merged, I agree with the unit test focus.  
  * However, issues like https://github.com/algorand/go-algorand/pull/3653#issuecomment-1083645448 make me feel we may revisit the topic in future iterations.